src/test/kotlin/cacophony/controlflow/generation/StructTest.kt

```diff
@@ -145,32 +132,298 @@ class StructTest {
+                ),
+            )
+
+        // when
+        val actualCFG = generateSimplifiedCFG(fDef)
+
+        // then
+        val virA = Register.VirtualRegister()
+        val virB = Register.VirtualRegister()
+        val virC = Register.VirtualRegister()
+
+        val expectedCFG =
+            singleWrappedFragmentCFG(fDef) {
+                "bodyEntry" does jump("assign b.a") { writeRegister(virA, integer(5)) }
+                "assign b.a" does jump("assign b.b") { writeRegister(virB, integer(7)) }
+                "assign b.b" does jump("reassign b.a") { writeRegister(virC, integer(1)) }
+                "reassign b.a" does jump("reassign b.b") { writeRegister(virB, integer(5)) }
+                "reassign b.b" does jump("assign res") { writeRegister(virC, integer(0)) }
+                "assign res" does jump("bodyExit") { writeRegister(getResultRegister(), registerUse(virB)) }
             }
 
         assertEquivalent(actualCFG, expectedCFG)
     }
+
+    // TODO: fix this test - accessing subfield makes whole structure land on stack
+    @Test
+    fun `access simple subfield in nested function`() {
+        // given
+
+        /*
+         * let f = [] -> Int => (
+         *  let x = {a = 7, b = true};
+         *  let g = [] -> Int => x.a = 3;
+         *  x.a
+         * );
+         */
+        val gDef =
+            intFunctionDefinition(
+                "g",
+                listOf(),
+                lvalueFieldRef(variableUse("x"), "a") assign lit(3),
+            )
+        val fDef =
+            intFunctionDefinition(
+                "f",
+                listOf(),
+                block(
+                    variableDeclaration("x", simpleStruct()),
+                    gDef,
+                    rvalueFieldRef(variableUse("x"), "a"),
+                ),
+            )
+
+        // when
+        val actualCFG = generateSimplifiedCFG(fDef)
+        val fCFG = actualCFG[fDef]!!
+        val gCFG = actualCFG[gDef]!!
+
+        // then
+
+        val expectedCFGf =
+            standaloneWrappedCFGFragment(fDef) {
+                "bodyEntry" does jump("assign b") { memoryAccess(registerUse(rbp) sub integer(16)) assign integer(7) }
+                "assign b" does jump("assign res") { memoryAccess(registerUse(rbp) sub integer(24)) assign integer(1) }
+                "assign res" does jump("bodyExit") { writeRegister(getResultRegister(), memoryAccess(registerUse(rbp) sub integer(16))) }
+            }
+
+        val expectedCFGg =
+            standaloneWrappedCFGFragment(gDef) {
+                "bodyEntry" does
+                    jump("bodyExit") {
+                        writeRegister(
+                            getResultRegister(),
+                            memoryAccess(memoryAccess(registerUse(rbp)) sub integer(16)) assign integer(3),
+                        )
+                    }
+            }
+
+        // TODO: expected should be different
```

Comment:
I think it is better to make expected CFG is as it should be, and we mark test as Disabled for now

Reply:
I am not sure how it should look atm, but I will update it before merge (after the other PR's merge as I mentioned)

Reply:
Resolved (by you)